### PR TITLE
feat: support dependencies json schema field

### PR DESCRIPTION
**Pull request by bytemain**
in JSON Schema Specification 7, it support `dependencies` in object type that can be used for data validate.

reference: https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#page-14

CleanShot 2023-05-25 at 11 07 35@2x

draft 7: https://json-schema.org/specification-links.html#draft-7

---

more info:
In JSON Schema 2019-09, `dependencies` has been split into `dependentSchemas` and `dependentRequired`

image

but currently we used the types from JSON Schema 7, so we can only support dependencies now.

### CODE REVIEW

**Comment by domoritz** [comment]
Thanks. I marked this as a draft.

**Comment by bytemain** [comment]
@domoritz Hello, This pull request is ready for review.

I'm sorry for updating this PR so late, I was a little busy recently.

**Comment by domoritz** [comment]
All good. Do you think it makes sense to update to a newer version of json schema? Also, is there some way we could use this functionality for some specific ts patterns rather than just for Jsdoc comments? Are the jadoc comments used anywhere else ow is this just a json schema thing?

**Comment by bytemain** [comment]
> 1. Do you think it makes sense to update to a newer version of json schema? 

I don't think there's a need to do that.

> 2. is there some way we could use this functionality for some specific ts patterns rather than just for Jsdoc comments.

Currently, it seem that TypeScript do not has a conception about `dependency`? In my use case, I'm using JSON Schemas to validate user input data, so this code already meets my needs.

**Comment by domoritz** [comment]
Thanks for the pull request. I think I would like to hear from someone else who wants this feature before adding it since there is no equivalent support in TS. Does that make sense?

**Comment by bytemain** [comment]
> Thanks for the pull request. I think I would like to hear from someone else who wants this feature before adding it since there is no equivalent support in TS. Does that make sense?

Ok, But in my opinion, the `dependentRequired` is the same kind of `title`/`format` in `BasicAnnotationsReader` class, it also read the jsdoc and save it to JSON Schemas:

https://github.com/vega/ts-json-schema-generator/blob/6b0e62e9e4794a28f60fa74ee332d03b026ca105/src/AnnotationsReader/BasicAnnotationsReader.ts#L10

it is a descriptor about this field.

So I think jsdoc is acceptable.

File: test/config.test.ts [hunk 1]
```diff
@@ -247,6 +247,15 @@ describe("config", () => {
             jsDoc: "extended",
         })
     );
+    it(
+        "jsdoc-dependentRequired",
+        assertSchema("jsdoc-dependentRequired", {
+            type: "Billing",
+            expose: "export",
+            topRef: true,
+            jsDoc: "extended",
```
Comment (domoritz) [inline test/config.test.ts:264]: Does the dependent feature work when the jsDoc reader is not enabled?

**Review by domoritz — COMMENTED** [review]
Good points. Let's merge then.

File: test/config.test.ts (hunk 1 above)
Comment (bytemain) [reply test/config.test.ts:264]: Yes, it does not rely on `jsDoc` config, should I support it(make it relying on the config)?

Comment (domoritz) [reply test/config.test.ts:264]: I’m not sure. What do you think are arguments either way ?

Comment (bytemain) [reply test/config.test.ts:264]: I think I should move my code to the `ExtendedAnnotationsReader` class, it is more reasonable.

But it seem that the `AnnotatedType` has no enough capability the handle dependency graph, I will look into it

Comment (domoritz) [reply test/config.test.ts:264]: Thanks. Let me know when you have revised.

**Comment by bytemain** [comment]
@domoritz I have moved the  `dependencies` related code to `ExtendedAnnotationsReader`.

I add a new function called `getTypeAnnotations` for `AnnotationsReader` interface:
```diff
export interface AnnotationsReader {
    getAnnotations(node: ts.Node): Annotations | undefined;
+    getTypeAnnotations?(type: BaseType): Annotations | undefined;
}
```

and then in the ExtendedAnnotationsReader, it can read things its need from the `ObjectType`, then create the `dependencies` we need.

File: src/AnnotationsReader/ExtendedAnnotationsReader.ts [hunk 2]
```diff
@@ -3,12 +3,14 @@ import ts from "typescript";
...
+import { ObjectType } from "../Type/ObjectType";
+import { StringMap } from "../Utils/StringMap";
 
-export class ExtendedAnnotationsReader extends BasicAnnotationsReader {
-    public constructor(
-        private typeChecker: ts.TypeChecker,
-        extraTags?: Set<string>
-    ) {
+export class ExtendedAnnotationsReader extends BasicAnnotationsReader implements AnnotationsReader {
+    public constructor(private typeChecker: ts.TypeChecker, extraTags?: Set<string>) {
```
Comment (bytemain) [inline src/AnnotationsReader/ExtendedAnnotationsReader.ts]: these style changes are made by `prettier`.

This repo's root has a prettier config file, so my vscode use prettier to format every file

**Comment by domoritz** [comment]
Can you resolve the conflicts @bytemain. Thanks for your patience with my review!

**Comment by bytemain** [comment]
> Can you resolve the conflicts @bytemain. Thanks for your patience with my review!

Done.

**Pull request closed without merging**